### ISSUE TITLE
Questions About DDPM

### ISSUE THREAD

**Issue by haoranD** [lucidrains/denoising-diffusion-pytorch#10]
Hi,

May I ask a question about the DDPM(based on implementation)? 

I am a bit confused about its training loss: loss = (**noise** - x_recon).abs().mean().

The **noise** is random noise? I am confused why it (the loss) is based on the random noise here? Forgive my foolishness and misunderstanding. Could you please explain a bit?

Thanks for your help.

**Comment by NielsRogge** [comment]
Hi,

So to my understanding, you want a neural network to learn (or approximate) a Gaussian conditional distribution. Any Gaussian distribution is characterised by 2 things: a mean (mu), and a standard deviation (sigma).

In the original DDPM paper, the authors decided to only make the neural network learn the mean (mu), not the standard deviation (they fixed the latter). After some math, they find that the mean (parametrised by theta, the parameters of the neural network), can be represented as follows:

<img width="856" alt="Screenshot 2022-04-08 at 12 01 42" src="https://user-images.githubusercontent.com/48327001/162413776-35504624-2930-45ba-a575-d546b32913bb.png">

So then they reframe this to make the neural network represent the Gaussian noise, and predict the true noise added at a particular time step t based on the input given at that time step.

**Comment by haoranD** [comment]
> Hi,
> 
> So to my understanding, you want a neural network to learn (or approximate) a Gaussian conditional distribution. Any Gaussian distribution is characterised by 2 things: a mean (mu), and a standard deviation (sigma).
> 
> In the original DDPM paper, the authors decided to only make the neural network learn the mean (mu), not the standard deviation (they fixed the latter). After some math, they find that the mean (parametrised by theta, the parameters of the neural network), can be represented as follows:
> 
> <img alt="Screenshot 2022-04-08 at 12 01 42" width="856" src="https://user-images.githubusercontent.com/48327001/162413776-35504624-2930-45ba-a575-d546b32913bb.png">
> 
> So then they reframe this to make the neural network represent the Gaussian noise, and predict the true noise added at a particular time step t based on the input given at that time step.

Many thanks for your help. Nice one! This makes sense to me.

**Comment by dfeng** [comment]
![](https://lilianweng.github.io/posts/2021-07-11-diffusion-models/DDPM-algo.png)

I think it's important to point out that `x_recon` is a function of the same `noise` term (you can see in the above algorithm that the `\epsilon` value is shared in the loss term). As your intuition rightly assumes, it wouldn't make sense for the loss to be just predicting random gaussian noise. So, the way I interpret it is: 

You use your `noise` (and `x_0`, which we have) to sample an `x_t` (that's the `q_sample` function).

https://github.com/lucidrains/denoising-diffusion-pytorch/blob/989f0fcb8e5f68ac851ccf213c993045512e7456/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L448-L454

Then your learned network `denoise_fn` takes in this sample `x_t` and `t`, and basically tries to recreate the original `noise`.

https://github.com/lucidrains/denoising-diffusion-pytorch/blob/989f0fcb8e5f68ac851ccf213c993045512e7456/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L460-L461

So really what you're doing is learning a model that can deconstruct an `x_t` into the signal (`x_0`) and the `noise`; or another way to put it, learn the "inverse function" of generating `x_t` from `x_0` and `noise`.

**Comment by haoranD** [comment]
> ![](https://camo.githubusercontent.com/0c8fc1a13a9938e30208188a06ea3de4b698108a8a5d80e0fe4defc07069180f/68747470733a2f2f6c696c69616e77656e672e6769746875622e696f2f706f7374732f323032312d30372d31312d646966667573696f6e2d6d6f64656c732f4444504d2d616c676f2e706e67)
> 
> I think it's important to point out that `x_recon` is a function of the same `noise` term (you can see in the above algorithm that the `\epsilon` value is shared in the loss term). As your intuition rightly assumes, it wouldn't make sense for the loss to be just predicting random gaussian noise. So, the way I interpret it is:
> 
> You use your `noise` (and `x_0`, which we have) to sample an `x_t` (that's the `q_sample` function).
> 
> https://github.com/lucidrains/denoising-diffusion-pytorch/blob/989f0fcb8e5f68ac851ccf213c993045512e7456/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L448-L454
> 
> Then your learned network `denoise_fn` takes in this sample `x_t` and `t`, and basically tries to recreate the original `noise`.
> 
> https://github.com/lucidrains/denoising-diffusion-pytorch/blob/989f0fcb8e5f68ac851ccf213c993045512e7456/denoising_diffusion_pytorch/denoising_diffusion_pytorch.py#L460-L461
> 
> So really what you're doing is learning a model that can deconstruct an `x_t` into the signal (`x_0`) and the `noise`; or another way to put it, learn the "inverse function" of generating `x_t` from `x_0` and `noise`.

Thanks for the sharing. This helps a lot.
